packages/model-viewer/src/features/annotation.ts

```diff
@@ -0,0 +1,177 @@
+    ModelViewerElement: T): Constructor<AnnotationInterface>&T => {
+  class AnnotationModelViewerElement extends ModelViewerElement {
+    private[$annotationRenderer] = new CSS2DRenderer();
+    private[$hotspotMap] = new Map();
+    private[$observer] = new MutationObserver((mutations) => {
+      mutations.forEach((mutation) => {
+        if (mutation.type === 'childList') {
+          mutation.addedNodes.forEach((node) => {
+            this[$addHotspot](node);
+          });
+          mutation.removedNodes.forEach((node) => {
+            this[$removeHotspot](node);
+          });
+        }
+      });
+    });
+
+    connectedCallback() {
+      super.connectedCallback();
+      const {domElement} = this[$annotationRenderer];
+      const {style} = domElement;
+      style.pointerEvents = 'none';
+      style.position = 'absolute';
+      style.top = '0';
+      this.shadowRoot!.querySelector('.container')!.appendChild(domElement);
+
+      for (let i = 0; i < this.children.length; ++i) {
+        this[$addHotspot](this.children[i]);
+      }
+
+      this[$observer].observe(this, {childList: true});
+    }
+
+    disconnectedCallback() {
+      super.disconnectedCallback();
+      this[$observer].disconnect();
+    }
+
+    updated(changedProperties: Map<string, any>) {
+      super.updated(changedProperties);
+    }
+
+    [$tick](time: number, delta: number) {
+      super[$tick](time, delta);
+      this[$updateHotspots]();
+      this[$annotationRenderer].render(this[$scene], this[$scene].activeCamera);
+    }
+
+    [$onResize](e: {width: number, height: number}) {
+      super[$onResize](e);
+      this[$annotationRenderer].setSize(e.width, e.height);
+    }
+
+    [$updateHotspots]() {
+      const {children} = this[$scene].pivot;
+      for (let i = 0, l = children.length; i < l; i++) {
+        const object = children[i];
+        if (object instanceof Hotspot) {
+          const view = this[$scene].activeCamera.position.clone();
+          view.sub(object.position);
+          if (view.dot(object.normal) < 0) {
+            object.element.classList.add('hide');
+          } else {
+            object.element.classList.remove('hide');
+          }
+        }
+      }
+    }
+
+    [$addHotspot](node: Node) {
+      if (!(node instanceof HTMLElement && node.slot.includes('hotspot')))
+        return;
+      const hotspot = new Hotspot({
+        name: node.slot,
+        position: node.dataset.position,
+        normal: node.dataset.normal
+      });
+      this[$hotspotMap].set(node, hotspot);
+      this[$scene].pivot.add(hotspot);
+    }
+
+    [$removeHotspot](node: Node) {
+      const hotspot = this[$hotspotMap].get(node);
+      if (!hotspot)
+        return;
+      this[$scene].pivot.remove(hotspot);
+      this[$hotspotMap].delete(node);
+    }
+
+    updateHotspot(config: HotspotConfiguration) {
+      const element = this.querySelector(`[slot="${config.name}"]`);
```

Comment:
Let's just keep a map of hotspots by name and validate with that. Names should be unique identifiers for hotspots anyway, so this map will be useful elsewhere.

In fact, you might not need to keep track of slotted elements at all.